### PR TITLE
Implement several Client snapshot methods

### PR DESCRIPTION
**Pull request by monoid**
+ allow_snapshot(hdfs_path)
  + disallow_snapshot(hdfs_path)
  + create_snapshot(hdfs_path, snapshotname=None)
  + delete_snapshot(hdfs_path, snapshotname)
  + rename_snapshot(hdfs_path, oldsnapshotname, snapshotname)

### CODE REVIEW

File: hdfs/client.py [hunk 1]
```diff
@@ -994,6 +999,75 @@ def checksum(self, hdfs_path):
...
+      or points to a normal file, an :class:`HdfsError` will be raised.
+    :param snapshotname snapshot name; if it does not exist, an
+      :class:`HdfsError` will be raised.
+    """
+    _logger.info('Deleting snapshot %r in %r.', snapshotname, hdfs_path)
+    hdfs_path = self.resolve(hdfs_path)
+    self._delete_snapshot(hdfs_path, snapshotname=snapshotname)
+
+  def rename_snapshot(self, hdfs_path, oldsnapshotname, snapshotname):
+    """Remove snapshot for a remote folder where it was allowed.
```
Comment (mtth) [inline hdfs/client.py:1055]: Should this be "Rename..."?

File: test/test_client.py [hunk 2]
```diff
@@ -1277,3 +1382,124 @@ def test_with_session(self):
...
+    self.client.allow_snapshot('foo')
+    self.client.create_snapshot('foo', 'mysnap')
+    self.client.delete_snapshot('foo', 'mysnap')
+
+  def test_create_snapshot_name(self):
+    self.client._mkdirs('foo')
+    self.client.allow_snapshot('foo')
+    try:
+      snapshot_path = self.client.create_snapshot('foo', 'mysnap')
+      assert snapshot_path.endswith('/foo/.snapshot/mysnap'), repr(snapshot_path)
```
Comment (mtth) [inline test/test_client.py:1438]: Can you switch to a testing utility (e.g. `ok_` or a more string-specific one) instead of `assert`?

**Review by mtth — COMMENTED** [review]
Thanks for the solid PR @monoid - nice documentation and tests!

CI is failing but it's unrelated to your feature (https://github.com/mtth/hdfs/issues/168). Were you able to test the new functionality on a working cluster?

File: hdfs/client.py (hunk 1 above)
Comment (monoid) [reply hdfs/client.py]: Indeed, thank you!  Fixed.

File: test/test_client.py (hunk 2 above)
Comment (monoid) [reply test/test_client.py]: Done (`assert_regex` seems to be the only option beyond `_ok`).

**Comment by monoid** [comment]
@mtth I ran tests in the docker with a simple docker-compose script:
```
# Run as:
#  docker-compose -f ./docker/docker-compose.yml run --rm testenv
version: '2.2'

services:
  hdfs:
    image: harisekhon/hadoop:2.8
  testenv:
    image: python:3.6
    depends_on:
      - hdfs
    volumes:
      - ../:/src:rw
    working_dir: /src
    command: bash -c 'pip3 install nose && pip3 install -e .[avro] && HDFSCLI_TEST_URL=http://hdfs:50070 nosetests'
```

Some unrelated tests failed because it is Hadoop 2.8 (GETTRASHROOT not available) and because ACL are not enabled out of box.  I'v failed to find a working 2.9 image.

**Comment by mtth** [comment]
Published as `2.6.0`.

**Comment by monoid** [comment]
@mtth thank you very much! :)

**Pull request merged**